include/mbedtls/ssl.h

```diff
@@ -2987,7 +2987,7 @@ void mbedtls_ssl_conf_dhm_min_bitlen( mbedtls_ssl_config *conf,
 #if defined(MBEDTLS_ECP_C)
 /**
  * \brief          Set the allowed curves in order of preference.
- *                 (Default: all defined curves.)
+ *                 (Default: all defined curves in order of decreasing size.)
```

Comment:
Just to be sure, in `ecp.c:ecp_supported_curves[]` the curves x25519 and x448 are at the end of the array and are not the "smallest". Thus is "in decreasing size" correct here ?

Reply:
Good point. Originally, `ecp_supported_curves` was in decreasing order, and the [comment at the top](https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.26.0/library/ecp.c#L505) still says “Curves are listed in order: largest curves first, and for a given size, fastest curves first”. This was true when the library only supported Weierstrass curves, but when we added Montgomery curves, we didn't pay attention and added them at the end of the list. So now the behavior is inconsistent with the documentation.

Changing the order is a behavior change that is in theory benign, but in practice it can have significant real-world consequences, such as triggering [this bug](https://github.com/ARMmbed/mbedtls/issues/4608), or suddenly no longer working with some buggy middlebox. That would not be acceptable in an TLS version, but it would be ok in a minor version.

@mpg @yanesca What do you think? Should we change the order of `ecp_supported_curves` to raise the priority of Montgomery curves in 2.27 to be in their order based on size? Would we put x448 (slow, not very popular) before x25519 (fast, very popular these days) — which could mean that x448 is preferred in the moderately common case where the peer accepts x448 but not Brainpool curves or P521?

Reply:
I think for 2.x we should change the order of curves in order to comply with what the documentation says even if that's sub-optimal: so x448 would go between the 512 and 384 bit curves and x25519 would go right after the 256-bit curves.

For 3.0 obviously we've changed that, and for 2.16 the Montgomery curves aren't there yet so no need to worry about their ordering.

Reply:
Since it is acceptable to do it in a minor version and it is not present in 2.16 I think it would make sense to align 2.x and 3.0 in this regard: easier to make backports later and I think it has a benefit for users as well if 2.x and 3.0 behave more similarly.

Reply:
I don't think aligning to 3.0 is an option, as 3.0 also removed some curves from the list, which we don't do in minor versions unless there's a pressing security reason. (As the added documentation in this PR says ^^)

Now I'm tempted to leave things as they are for now (keep the inconsistent ordering and the vague hence not wrong documentation), and before 2.28, introduce a new SSL preset `MBEDTLS_SSL_PRESET_NEXT` (or any better name) that would match the defaults of 3.0, which we would advertise for its performance benefits.

Reply:
I'm certainly not comfortable with changing the order without providing an easy way to keep the old order, which would mean a new preset.

If Curve25519 becomes the default, we'll run into a practical problem that this PR will need to be rebased on top of https://github.com/ARMmbed/mbedtls/pull/4623. I'd rather not add lead time here since we're close to the release code freeze.

Proposal: in this PR, only tweak the documentation here to say that Montgomery curves are last in the order and this may change in a future version. Then, after 2.27, we'll create a new preset and tweak the order in the strongest-curve-first preset. I've filed https://github.com/ARMmbed/mbedtls/issues/4698, with some details TBD. Sound good?

Reply:
Sounds good to me!

Reply:
I am ok with that as well.

Reply:
Done.